dftimewolf/lib/args_validator.py

```diff
@@ -0,0 +1,226 @@
+
+import abc
+import ipaddress
+import re
+
+from typing import Any, Dict, Optional
+
+from dftimewolf.lib import errors
+
+
+class AbstractValidator(abc.ABC):
+  """Base class for validator objects."""
+
+  name: str = None  # type: ignore
+
+  @abc.abstractmethod
+  def __init__(self) -> None:
+    """Initialise."""
+
+  @abc.abstractmethod
+  def Validate(
+      self, operand: Any, validator_params: Optional[Dict[str, Any]]) -> None:
+    """Validate the parameter.
+
+    Subclasses that override this should raise errors.RecipeArgsValidatorError
+    on validation failures, with a string describing the problem."""
+
+
+class CommaSeparatedValidator(AbstractValidator):
+  """Subclass of AbstractValidator that allows for comma separated strings.
+
+  Subclasses that override this should implement ValidateSingle instead of
+  Validate."""
+
+  def Validate(self,
+               operand: str,
+               validator_params: Dict[str, Any]) -> None:
+    """Split the string by commas if validator_params['comma_separated'] == True
+    and validate each component in ValidateSingle."""
+
+    if 'comma_separated' not in validator_params:
+      raise errors.RecipeArgsValidatorError(
+          'Missing validator parameter: comma_separated')
+
+    operands = []
+    if validator_params['comma_separated']:
+      operands = operand.split(',')
+    else:
+      operands.append(operand)
+
+    for op in operands:
+      self.ValidateSingle(op, validator_params)
+
+  @abc.abstractmethod
+  def ValidateSingle(
+      self, operand: str, validator_params: Optional[Dict[str, Any]]) -> None:
+    """Validate a single operand from a comma separated list."""
+
+
+class DefaultValidator(AbstractValidator):
+  """The default validator always returns true."""
+
+  def __init__(self) -> None:
+    """Initialise."""
+    super().__init__()
+    self.name = 'default'
+
+  def Validate(self,
+               operand: Any,
+               validator_params: Optional[Dict[str, Any]]) -> None:
+    """Never raises an errors.RecipeArgsValidatorError."""
+
+
+class AWSRegionValidator(AbstractValidator):
+  """Validates a correct AWS region."""
+
+  # pylint: disable=line-too-long
+  # Source: curl -s https://ip-ranges.amazonaws.com/ip-ranges.json | jq -r '.prefixes[] | select(.service == "EC2") | .region' ip-ranges.json  | sort | uniq
+  # Fetched 2023-01-15
+  # pylint: enable=line-too-long
+  _regions = [
+    'af-south-1', 'ap-east-1', 'ap-northeast-1', 'ap-northeast-2',
+    'ap-northeast-3', 'ap-south-1', 'ap-south-2', 'ap-southeast-1',
+    'ap-southeast-2', 'ap-southeast-3', 'ap-southeast-4', 'ap-southeast-6',
+    'ca-central-1', 'ca-west-1', 'cn-north-1', 'cn-northwest-1', 'eu-central-1',
+    'eu-central-2', 'eu-north-1', 'eu-south-1', 'eu-south-2', 'eu-west-1',
+    'eu-west-2', 'eu-west-3', 'il-central-1', 'me-central-1', 'me-south-1',
+    'sa-east-1', 'us-east-1', 'us-east-2', 'us-gov-east-1', 'us-gov-west-1',
+    'us-west-1', 'us-west-2'
+  ]
```

Comment:
Make this a set, so that the `in` operation is faster.

```suggestion
  _regions = {
    'af-south-1', 'ap-east-1', 'ap-northeast-1', 'ap-northeast-2',
    'ap-northeast-3', 'ap-south-1', 'ap-south-2', 'ap-southeast-1',
    'ap-southeast-2', 'ap-southeast-3', 'ap-southeast-4', 'ap-southeast-6',
    'ca-central-1', 'ca-west-1', 'cn-north-1', 'cn-northwest-1', 'eu-central-1',
    'eu-central-2', 'eu-north-1', 'eu-south-1', 'eu-south-2', 'eu-west-1',
    'eu-west-2', 'eu-west-3', 'il-central-1', 'me-central-1', 'me-south-1',
    'sa-east-1', 'us-east-1', 'us-east-2', 'us-gov-east-1', 'us-gov-west-1',
    'us-west-1', 'us-west-2'
  }
```

Reply:
(same for other similar validators)

Reply:
Done.